homeassistant/components/tplink/__init__.py

```diff
@@ -130,3 +165,61 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         hass.data[DOMAIN].clear()
 
     return unload_ok
+
+
+class SmartPlugDataUpdateCoordinator(DataUpdateCoordinator):
+    """DataUpdateCoordinator to gather data for specific SmartPlug."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        smartplug: SmartPlug,
+    ) -> None:
+        """Initialize DataUpdateCoordinator to gather data for specific SmartPlug."""
+        self.smartplug = smartplug
+
+        update_interval = timedelta(seconds=30)
+        super().__init__(
+            hass, _LOGGER, name=smartplug.alias, update_interval=update_interval
+        )
+
+    async def _async_update_data(self) -> dict:
+        """Fetch all device and sensor data from api."""
+        info = self.smartplug.sys_info
+        data = {}
+        data[CONF_HOST] = self.smartplug.host
+        data[CONF_MAC] = info["mac"]
+        data[CONF_MODEL] = info["model"]
+        data[CONF_SW_VERSION] = info["sw_ver"]
+        if self.smartplug.context is None:
+            data[CONF_ALIAS] = info["alias"]
+            data[CONF_DEVICE_ID] = info["mac"]
+            data[CONF_STATE] = self.smartplug.state == self.smartplug.SWITCH_STATE_ON
+        else:
+            plug_from_context = next(
+                c
+                for c in self.smartplug.sys_info["children"]
+                if c["id"] == self.smartplug.context
+            )
+            data[CONF_ALIAS] = plug_from_context["alias"]
+            data[CONF_DEVICE_ID] = self.smartplug.context
+            data[CONF_STATE] = plug_from_context["state"] == 1
+        if self.smartplug.has_emeter:
+            emeter_readings = self.smartplug.get_emeter_realtime()
+            data[CONF_EMETER_PARAMS] = {
+                ATTR_CURRENT_POWER_W: round(float(emeter_readings["power"]), 2),
+                ATTR_TOTAL_ENERGY_KWH: round(float(emeter_readings["total"]), 3),
+                ATTR_VOLTAGE: round(float(emeter_readings["voltage"]), 1),
+                ATTR_CURRENT_A: round(float(emeter_readings["current"]), 2),
+                ATTR_LAST_RESET: {ATTR_TOTAL_ENERGY_KWH: utc_from_timestamp(0)},
+            }
+            emeter_statics = self.smartplug.get_emeter_daily()
+            if emeter_statics.get(int(time.strftime("%e"))):
+                data[CONF_EMETER_PARAMS][ATTR_LAST_RESET][
```

Comment:
Last reset should be an isoformatted datetime string

Reply:
looks now like
![image](https://user-images.githubusercontent.com/35783820/127481663-5af47b89-52b7-4497-a750-a0014e14e65e.png)